packages/hash/shared/src/ProsemirrorManager.ts

```diff
@@ -0,0 +1,530 @@
+          },
+          [this.schema.nodes[targetComponentId]!.create({}, content)],
+        ),
+      ),
+    ]);
+  }
+
+  private assertBlockDefined(componentId: string) {
+    if (!this.schema.nodes[componentId]) {
+      throw new Error("Block must already be defined");
+    }
+  }
+
+  async loadPage(currentState: EditorState<Schema>, entities: BlockEntity[]) {
+    const store = createEntityStore(
+      entities,
+      entityStorePluginState(currentState).store.draft,
+    );
+
+    const newNodes = await Promise.all(
+      entities.map(async (blockEntity) => {
+        const draftEntity = mustGetDraftEntityByEntityId(
+          store.draft,
+          blockEntity.entityId,
+        );
+
+        await this.defineBlockByComponentId(blockEntity.properties.componentId);
+
+        return this.renderBlock(
+          blockEntity.properties.componentId,
+          store,
+          draftEntity.draftId,
+        );
+      }),
+    );
+
+    const { tr } = currentState;
+
+    addEntityStoreAction(currentState, tr, { type: "store", payload: store });
+
+    tr.replaceWith(0, currentState.doc.content.size, newNodes);
+
+    return tr;
+  }
+
+  /**
+   * @todo consider removing the old block from the entity store
+   */
+  async replaceNode(
+    draftBlockId: string,
+    targetComponentId: string,
+    targetVariant: BlockVariant,
+    node: ProsemirrorNode<Schema>,
+    pos: number,
+  ) {
+    const { view } = this;
+
+    // @todo consider separating the logic that uses a view to something else
+    if (!view) {
+      throw new Error("Cannot trigger replaceNodeWithRemoteBlock without view");
+    }
+
+    const [tr, newNode] = await this.createBlock(
+      targetComponentId,
+      draftBlockId,
+      targetVariant,
+    );
+
+    tr.replaceRangeWith(pos, pos + node.nodeSize, newNode);
+    view.dispatch(tr);
+  }
+
+  /**
+   * @todo consider removing the old block from the entity store
+   */
+  async deleteNode(node: ProsemirrorNode<Schema>, pos: number) {
+    const { view } = this;
+
+    if (!view) {
+      throw new Error("Cannot trigger deleteNode without view");
+    }
+
+    const { tr } = view.state;
+
+    tr.delete(pos, pos + node.nodeSize);
+    view.dispatch(tr);
+  }
+
+  async createBlock(
+    targetComponentId: string,
+    draftBlockId: string | null,
+    targetVariant: BlockVariant,
+  ) {
+    if (!this.view) {
+      throw new Error("Cannot trigger createBlock without view");
+    }
+
+    await this.defineBlockByComponentId(targetComponentId);
+
+    const { tr } = this.view.state;
+
+    const entityProperties = targetVariant?.properties ?? {};
+    const entityStoreState = entityStorePluginState(this.view.state);
+    const blockEntity = draftBlockId
+      ? entityStoreState.store.draft[draftBlockId]
+      : null;
+
+    if (
+      (draftBlockId && !blockEntity) ||
+      (blockEntity && !isDraftBlockEntity(blockEntity))
+    ) {
+      throw new Error("draft id does not belong to a block");
+    }
```

Comment:
I feel breaking this into 2 conditionals will make errors here easier to debug

```suggestion
   if (draftBlockId && !blockEntity){
    throw new Error("draft id is not present in entity store")
   }
   
    if (blockEntity && !isDraftBlockEntity(blockEntity)) {
      throw new Error("draft id does not belong to a block");
    }
```

Reply:
I think there's a formatting issue with the suggestion, so did this manually, but the suggestion has been accepted in [`d382961` (#872)](https://github.com/hashintel/hash/pull/872/commits/d382961d9f1ea3e7b41b532cdc38411851d9cf39)